File: Week 5/Day 5/HAKATHON/backend/seed.js
// Seed the database with sample reports on first run (mirrors the frontend demo).

const db = require('./db');
const reports = require('./reports');

const SEED = [
  { id: 'MTF-2026-K2N7', cat: 'roads', ward: 'Kasarani', landmark: 'Near Kasarani Stadium gate B',
    desc: 'Huge pothole covering half the road on Kasarani–Mwiki road. Boda riders are swerving into oncoming traffic to avoid it.',
    name: 'James Mwangi', phone: '0712345678', status: 4, ageDays: 12 },
  { id: 'MTF-2026-P9Q4', cat: 'water', ward: "Kibera (Sarang'ombe)", landmark: 'Behind Toi Market',
    desc: 'Water pipe burst three days ago. Water is flowing into the road, and we have had no supply at home since yesterday.',
    name: 'Achieng O.', phone: '0722111222', status: 3, ageDays: 5 },
  { id: 'MTF-2026-D4F8', cat: 'lights', ward: 'Westlands', landmark: 'Parklands 5th Avenue',
    desc: 'Streetlights along 5th Avenue have been off for two weeks. The stretch is now unsafe at night, especially for women walking from work.',
    name: 'Fatuma Ali', phone: '0733444555', status: 2, ageDays: 3 },
  { id: 'MTF-2026-M6R2', cat: 'garbage', ward: 'Embakasi East', landmark: 'Nyayo Estate Phase 2 gate',
    desc: 'Garbage has not been collected for three weeks. The heap is now blocking part of the road and smells terrible.',
    name: 'Peter Njoroge', phone: '0744666777', status: 1, ageDays: 2 },
  { id: 'MTF-2026-W8T5', cat: 'drainage', ward: 'Mvita, Mombasa', landmark: 'Near Makadara Mosque',
    desc: 'Blocked drainage on Abdel Nasser Road. Every light rain floods the shops along the road.',
    name: 'Swaleh Omar', phone: '0755888999', status: 0, ageDays: 1 }
];

async function ensureSeed() {
  if (reports.count() > 0) return; // already has data
  const now = Date.now();
  for (const s of SEED) {
    const created = new Date(now - s.ageDays * 86400000).toISOString();
    db.run(
      `INSERT INTO reports (id, cat, ward, landmark, description, name, phone, photo, status, created, updated)
       VALUES (?, ?, ?, ?, ?, ?, ?, NULL, ?, ?, ?)`,
      [s.id, s.cat, s.ward, s.landmark, s.desc, s.name, s.phone, s.status, created, created]
    );
    // build a simple status history up to the current status
    for (let i = 0; i <= s.status; i++) {
      db.run('INSERT INTO status_history (report_id, status, note, at) VALUES (?, ?, ?, ?)',
        [s.id, i, i === 0 ? 'Received by Mtaafix' : null, created]);
    }
  }
  console.log(`Seeded ${SEED.length} sample reports.`);
}

module.exports = { ensureSeed };

// allow `npm run seed`
if (require.main === module) {
  db.init().then(ensureSeed).then(() => process.exit(0));
}

File: Week 5/Day 5/Hackathon/mtaafix/backend/seed.py
"""Seed rulers, a couple of residents, and sample reports on first run."""

import os
from datetime import datetime, timedelta, timezone

import db
import auth
import reports
import residents
from constants import AREAS

# One area ruler per area (demo passwords), plus a super-admin ruler.
RULERS = [
    {"username": "kasarani-ruler", "name": "Kasarani Area Ruler", "area": "kasarani"},
    {"username": "kibera-ruler",   "name": "Kibera Area Ruler",   "area": "kibera"},
    {"username": "westlands-ruler", "name": "Westlands Area Ruler", "area": "westlands"},
    {"username": "embakasi-ruler", "name": "Embakasi Area Ruler", "area": "embakasi"},
    {"username": "mvita-ruler",    "name": "Mvita Area Ruler",    "area": "mvita"},
]

SEED_REPORTS = [
    {"cat": "roads", "area": "kasarani", "ward": "Kasarani", "landmark": "Near Kasarani Stadium gate B",
     "desc": "Pothole repair works on Kasarani-Mwiki road begin this week. Expect diversions between 8am and 4pm.",
     "author": "kasarani-ruler", "status": 3, "ageDays": 4},
    {"cat": "water", "area": "kibera", "ward": "Sarang'ombe", "landmark": "Behind Toi Market",
     "desc": "Burst water pipe behind Toi Market has been reported to the utility. Repair team dispatched.",
     "author": "kibera-ruler", "status": 2, "ageDays": 2},
    {"cat": "lights", "area": "westlands", "ward": "Parklands", "landmark": "5th Avenue",
     "desc": "Streetlights along 5th Avenue will be restored on Friday. Please stay alert at night until then.",
     "author": "westlands-ruler", "status": 1, "ageDays": 1},
]


def ensure_seed():
    if auth.count_rulers() == 0:
        admin_user = os.environ.get("ADMIN_USER", "admin")
        admin_pass = os.environ.get("ADMIN_PASSWORD", "admin123")
        auth.create_ruler(admin_user, admin_pass, "Chief Administrator", role="admin", area=None)
        print(f'Created super-admin ruler "{admin_user}".')
        if not os.environ.get("ADMIN_PASSWORD"):
            print('  WARNING: Using default admin password "admin123" - set ADMIN_PASSWORD and change it!')
        for r in RULERS:
            auth.create_ruler(r["username"], f"{r['area']}123", r["name"], role="ruler", area=r["area"])
            print(f'Created area ruler "{r["username"]}".')
        print('  WARNING: Area ruler demo passwords are "<areaId>123" - change them in production!')

    if residents.count() == 0:
        residents.register_or_login("0712345678", name="Jane Resident", area="kasarani")
        residents.register_or_login("0722111222", name="Peter Resident", area=None)
        print("Seeded 2 demo residents.")

    if reports.count() == 0:
        now = datetime.now(timezone.utc)
        for s in SEED_REPORTS:
            ruler = auth.find_ruler(s["author"])
            author = {"id": ruler["id"], "name": ruler["name"]}
            created = (now - timedelta(days=s["ageDays"])).isoformat()
            rid = reports.gen_id()
            db.run(
                """INSERT INTO reports (id, cat, area, ward, landmark, description,
                       author_id, author_name, photo, status, created, updated)
                   VALUES (?, ?, ?, ?, ?, ?, ?, ?, NULL, ?, ?, ?)""",
                (rid, s["cat"], s["area"], s["ward"], s["landmark"], s["desc"],
                 author["id"], author["name"], s["status"], created, created))
            for i in range(s["status"] + 1):
                db.run("INSERT INTO status_history (report_id, status, note, at) VALUES (?, ?, ?, ?)",
                       (rid, i, ("Posted by " + author["name"]) if i == 0 else None, created))
        print(f"Seeded {len(SEED_REPORTS)} sample reports.")


if __name__ == "__main__":
    db.init()
    ensure_seed()